File: demo/midgard-sdk/src/types/contracts/retired-operators.ts
import { Data } from "@lucid-evolution/lucid";
import { POSIXTimeSchema } from "./common.js";

export const DatumSchema = Data.Object({
  commitmentTime: Data.Nullable(POSIXTimeSchema),
});
export type Datum = Data.Static<typeof DatumSchema>;
export const Datum = DatumSchema as unknown as Datum;

export const MintRedeemerSchema = Data.Enum([
  Data.Literal("Init"),
  Data.Literal("Deinit"),
  Data.Object({
    RetireOperator: Data.Object({
      newRetireOperatorKey: Data.Bytes(),
      hubOracleRefInputIndex: Data.Integer(),
      retiredOperatorAppendedNodeOutputIndex: Data.Integer(),
      retiredOperatorAnchorNodeOutputIndex: Data.Integer(),
      activeOperatorsRedeemerIndex: Data.Integer(),
    }),
  }),
  Data.Object({
    RecoverOperatorBond: Data.Object({
      retiredOperatorKey: Data.Bytes(),
      removedNodeInputIndex: Data.Integer(),
      anchorNodeInputIndex: Data.Integer(),
    }),
  }),
  Data.Object({
    RemoveOperatorSlashBond: Data.Object({
      slashedRetiredOperatorKey: Data.Bytes(),
      hubOracleRefInputIndex: Data.Integer(),
      retiredOperatorSlashedNodeInputIndex: Data.Integer(),
      retiredOperatorAnchorNodeInputIndex: Data.Integer(),
      state_queueRedeemerIndex: Data.Integer(),
    }),
  }),
]);
export type MintRedeemer = Data.Static<typeof MintRedeemerSchema>;
export const MintRedeemer = MintRedeemerSchema as unknown as MintRedeemer;

File: offchain/src/Midgard/Types/RetiredOperators.hs
{-# LANGUAGE TemplateHaskell #-}

module Midgard.Types.RetiredOperators (Datum, NodeData (..), MintRedeemer (..), rootAssetName, nodeAssetNamePrefix, nodeAssetNamePrefixLen) where

import Data.ByteString (ByteString)
import Data.ByteString.Char8 qualified as BS8
import GHC.Generics (Generic)

import Cardano.Api qualified as C
import PlutusLedgerApi.V3 (POSIXTime, PubKeyHash)
import PlutusTx.Blueprint (HasBlueprintDefinition, definitionRef)
import PlutusTx.Blueprint.TH (makeIsDataSchemaIndexed)

import Midgard.Types.LinkedList qualified as LinkedList
import PlutusLedgerApi.Common
import Ply (PlyArg)

rootAssetName :: C.AssetName
rootAssetName = C.UnsafeAssetName $ BS8.pack "MIDGARD_RETIRED_OPERATORS"

nodeAssetNamePrefix :: ByteString
nodeAssetNamePrefix = BS8.pack "MRET"

nodeAssetNamePrefixLen :: Int
nodeAssetNamePrefixLen = BS8.length nodeAssetNamePrefix

newtype NodeData = NodeData
  { bondUnlockTime :: Maybe POSIXTime
  }
  deriving stock (Eq, Show, Generic)
  deriving anyclass (HasBlueprintDefinition)

$( makeIsDataSchemaIndexed
     ''NodeData
     [ ('NodeData, 0)
     ]
 )

type Datum = LinkedList.Element BuiltinByteString NodeData

data MintRedeemer
  = Init {outputIndex :: Integer}
  | Deinit {inputIndex :: Integer}
  | RetireOperator
      { newRetiredOperatorKey :: PubKeyHash
      , bondUnlockTime :: Maybe POSIXTime
      , hubOracleRefInputIndex :: Integer
      , retiredOperatorAnchorElementInputIndex :: Integer
      , retiredOperatorAnchorElementOutputIndex :: Integer
      , retiredOperatorInsertedNodeOutputIndex :: Integer
      , activeOperatorsRedeemerIndex :: Integer
      }
  | RecoverOperatorBond
      { retiredOperatorKey :: PubKeyHash
      , retiredOperatorAnchorElementInputIndex :: Integer
      , retiredOperatorRemovedNodeInputIndex :: Integer
      , retiredOperatorAnchorElementOutputIndex :: Integer
      }
  | RemoveOperatorBadState
      { slashedRetiredOperatorKey :: PubKeyHash
      , hubOracleRefInputIndex :: Integer
      , retiredOperatorAnchorElementInputIndex :: Integer
      , retiredOperatorSlashedNodeInputIndex :: Integer
      , retiredOperatorAnchorElementOutputIndex :: Integer
      , stateQueueRedeemerIndex :: Integer
      }
  | RemoveOperatorBadSettlement
      { slashedRetiredOperatorKey :: PubKeyHash
      , hubOracleRefInputIndex :: Integer
      , retiredOperatorAnchorElementInputIndex :: Integer
      , retiredOperatorSlashedNodeInputIndex :: Integer
      , retiredOperatorAnchorElementOutputIndex :: Integer
      , settlementInputIndex :: Integer
      , settlementRedeemerIndex :: Integer
      }
  deriving stock (Eq, Show, Generic)
  deriving anyclass (HasBlueprintDefinition)

$( makeIsDataSchemaIndexed
     ''MintRedeemer
     [ ('Init, 0)
     , ('Deinit, 1)
     , ('RetireOperator, 2)
     , ('RecoverOperatorBond, 3)
     , ('RemoveOperatorBadState, 4)
     , ('RemoveOperatorBadSettlement, 5)
     ]
 )

instance PlyArg MintRedeemer
